src/dawn/SIR/SIR.proto

```diff
@@ -0,0 +1,627 @@
+  string callee = 2;            // Name of the called stencil (i.e callee)
+  repeated Field arguments = 3; // Fields passed as arguments during the stencil call
+}
+
+/*===------------------------------------------------------------------------------------------===*\
+ *     AST
+\*===------------------------------------------------------------------------------------------===*/
+
+// @brief This is a union of all the statements
+// @ingroup sir_proto
+message Stmt {
+  oneof stmt {
+    BlockStmt block_stmt = 1;
+    ExprStmt expr_stmt = 2;
+    ReturnStmt return_stmt = 3;
+    VarDeclStmt var_decl_stmt = 4;
+    StencilCallDeclStmt stencil_call_decl_stmt = 5;
+    VerticalRegionDeclStmt vertical_region_decl_stmt = 6;
+    BoundaryConditionDeclStmt boundary_condition_decl_stmt = 7;
+    IfStmt if_stmt = 8;
+  }
+}
+
+// @brief This is a union of all the expressions
+// @ingroup sir_proto
+message Expr {
+  oneof expr {
+    UnaryOperator unary_operator = 1;
+    BinaryOperator binary_operator = 2;
+    AssignmentExpr assignment_expr = 3;
+    TernaryOperator ternary_operator = 4;
+    FunCallExpr fun_call_expr = 5;
+    StencilFunCallExpr stencil_fun_call_expr = 6;
+    StencilFunArgExpr stencil_fun_arg_expr = 7;
+    VarAccessExpr var_access_expr = 8;
+    FieldAccessExpr field_access_expr = 9;
+    LiteralAccessExpr literal_access_expr = 10;
+  }
+}
+
+// @brief A block of `N` statetements
+//
+// Synopsis
+// @code{.cpp}
+// {
+//  statement_1
+//  statement_2
+//  ...
+//  statement_N
+// }
+// @endcode
+// @ingroup sir_proto
+message BlockStmt {
+  repeated Stmt statements = 1; // List of statements
+  SourceLocation loc = 2;       // Source location
+}
+
+// @brief A statement wrapping an expression
+//
+// @ingroup sir_proto
+message ExprStmt {
+  Expr expr = 1;          // Expression
+  SourceLocation loc = 2; // Source location
+}
+
+// @brief Return statement
+//
+// @ingroup sir_proto
+message ReturnStmt {
+  Expr expr = 1;          // Expression to return
+  SourceLocation loc = 2; // Source location
+}
+
+// @brief Declaration of a variable
+//
+// Synopsis
+// @code{.cpp}
+// Type name[dimension] op init_list
+// @endcode
+//
+// @b Example
+// In the following C++ expression:
+//
+// @code{.cpp}
+//  double foo[2] = {2.3, 5.3};
+// @endcode
+//
+// the members of `VarDeclStmt` would be:
+//
+// @code{.cpp}
+//  - type: Float
+//  - name: "foo"
+//  - dimension: 2
+//  - op: "="
+//  - init_list: {LiteralAccessExpr{"2.3"}, LiteralAccessExpr{"5.3"}}
+// @endcode
+//
+// @ingroup sir_proto
+message VarDeclStmt {
+  Type type = 1;       // Type of the variable
+  string name = 2;     // Name of the variable
+  int32 dimension = 3; // Dimension of the array or 0 for variables
+  string op = 4;       // Operation used for initialization
+
+  // List of expression used for array initialization or just 1 element for variable initialization
+  repeated Expr init_list = 5;
```

Comment:
how do we ensure consistency of the SIR? 
In this particular case, we need to ensure that the repetition number equals the dimension. Is it possible to enforce this in protobuf? If not how to we do it, do we need to develop validity scripts? (would be nice not to have to...)

Reply:
Consistency checking has to be done on the client side. Meaning you either make these checks *before* you write the SIR or directly *after* parsing. I would argue that the consumer of the SIR (for example Dawn) needs to perform the consistency checks. Protobuf itself does not support any checks afaik besides adhering to the provided scheme. In this particular case however it doesn't need to match the dimensionality as one could write
```
doube a[5] = {0};
``` 
which initializes everything to 0.

Reply:
so what happens if in the VarDeclStmt message one has a `dimension = 3` and a `init_lint = {0,0,0,0}`

Reply:
This is illformed C++ code and will be rejected by clang in gtclang. I think this should be checked by the one who builds the SIR i.e the frontend DSL.